### PR TITLE
Introduced "workflowsToDeleteMetadata" gauge metric [BA-6250]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala [hunk 1]
```diff
@@ -81,6 +81,16 @@ trait WorkflowMetadataSummaryEntryComponent {
...
 
+  val countRootWorkflowIdsByArchiveStatusAndEndedOnOrBeforeThresholdTimestamp = Compiled(
+    (metadataArchiveStatus: Rep[Option[String]], workflowEndTimestampThreshold: Rep[Timestamp]) => {
+      (for {
+        summary <- workflowMetadataSummaryEntries
+        if summary.rootWorkflowExecutionUuid.isEmpty && summary.parentWorkflowExecutionUuid.isEmpty // is root workflow entry
+        if summary.metadataArchiveStatus === metadataArchiveStatus
+        if summary.endTimestamp <= workflowEndTimestampThreshold
+      } yield summary.workflowExecutionUuid).length
+    })
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: There is already an index on metadata archive status but currently no index on end timestamp. So it seems that if there are a large number of Archived workflows this query could do a lot of scanning, not unlike what we saw with the backed up summarization queue.

Comment (mcovarr) [reply database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: ...although I'm not sure how much an end timestamp index would help with `<=` 🙁 . Maybe a config switch to turn off calculating this metric if we know things are backed up (like yesterday)?

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala [hunk 2]
```diff
@@ -45,12 +50,27 @@ class DeleteMetadataActor(metadataDeletionConfig: MetadataDeletionConfig) extend
...
+    // we need the dedicated `count` query here because the result of the `queryRootWorkflowSummaryEntriesByArchiveStatusAndOlderThanTimestamp`
+    // query is upper-bounded by `batchSize` parameter, which is undesired here
+    countRootWorkflowSummaryEntriesByArchiveStatusAndOlderThanTimestamp(
+      MetadataArchiveStatus.toDatabaseValue(Archived),
+      currentTimestampMinusDelay
+    ) onComplete {
+      case Success(numToDelete) =>
+        sendGauge(workflowsToDeleteMetadataMetricPath, numToDelete.toLong, InstrumentationPrefixes.ServicesPrefix)
+      case Failure(ex) =>
+        log.warning(s"Cannot send workflowsToDeleteMetadata gauge metric value: unable to query number of workflows-to-delete-metadata from the database: ${ex.getMessage}")
```
Comment (mcovarr) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala]: TOL This really shouldn't happen; perhaps `log.error` with the Throwable to get the stack trace?

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala (hunk 1 above)
Comment (cjllanwarne) [reply database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: I think it's fine that the metric is potentially slow, my request would be that it should not happen in serial with the actual processing of metadata deletion.

I would say:

* If the number of metadata entries returned by the deletion processing is less than the maximum batch size, we know exactly how many there are needing deletion so we don't need to re-run this.
* If the number of metadata entries returned by the deletion processing query is equal to the maximum batch size, we should activate a parallel action to count how many there are, and continue polling at some interval until the backlog falls back to zero.

File: hybridCarboniteMetadataService/src/test/scala/cromwell/services/metadata/hybridcarbonite/WorkflowsToDeleteMetadataMetricHelperActorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+class WorkflowsToDeleteMetadataMetricHelperActorSpec extends TestKitSuite with FlatSpecLike with ImplicitSender {
+
+  it should "send MetricActorFreed to DeleteMetadataActor after restart" in {
+    val deleteMetadataActor = TestProbe()
+    val metricActor = system.actorOf(Props(new WorkflowsToDeleteMetadataMetricHelperActor(deleteMetadataActor.ref, TestProbe().ref) {
+      override def receive: Receive = {
+        case SendPrecalculatedMetric(_) =>
+          throw new RuntimeException("TEST: runtime exception")
+      }
```
Comment (mcovarr) [inline hybridCarboniteMetadataService/src/test/scala/cromwell/services/metadata/hybridcarbonite/WorkflowsToDeleteMetadataMetricHelperActorSpec.scala]: ooh and a test for the 💯: 🥇

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/WorkflowsToDeleteMetadataMetricHelperActor.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  implicit val ec = context.dispatcher
+
+  override def postRestart(reason: Throwable): Unit = {
+    super.postRestart(reason)
+    metadataDeletionActor ! MetricActorFreed
+  }
+
+  override def receive: Receive = {
+    case SendPrecalculatedMetric(workflowsToDeleteMetadata) =>
```
Comment (cjllanwarne) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/WorkflowsToDeleteMetadataMetricHelperActor.scala]: I would give these messages more specific names - it took me a while to realize that they were specific messages for "archived workflows deletion queue" metrics

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala [hunk 5]
```diff
@@ -2,29 +2,38 @@ package cromwell.services.metadata.hybridcarbonite
...
 
   metadataDeletionConfig.intervalOpt.map { interval =>
     context.system.scheduler.schedule(5.minutes, interval, self, DeleteMetadataAction)
   }
 
-  override def receive: Receive = {
+  val workflowsToDeleteMetadataMetricHelperActor =
+    context.actorOf(WorkflowsToDeleteMetadataMetricHelperActor.props(self, serviceRegistryActor))
+
+  override def receive: Receive = receive(isMetricActorFree = true)
```
[the remark this answers is not in the document]
Comment (grsterin) [reply hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala]: That was my initial implementation, but Idea suggested to replace it with `become` behavior. And actually it looks more in the spirit of Akka than var.

[the remark this answers is not in the document]
Comment (cjllanwarne) [reply hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala]: hmm, I guess I disagree with Idea in this case, but I won't push it! I suspect the most "akka"y way to do this would be to have two separately coded methods to switch between, rather than methods which get generated ad hoc using a boolean parameter. But there's so much shared between them that that seems like huge overkill here.

TL;DR: You should stick with whichever you prefer!

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala [hunk 6]
```diff
@@ -34,6 +43,18 @@ class DeleteMetadataActor(metadataDeletionConfig: MetadataDeletionConfig) extend
       )
       workflowIdsForMetadataDeletionFuture onComplete {
         case Success(workflowIds) =>
+          if (isMetricActorFree) {
```
[the remark this answers is not in the document]
Comment (grsterin) [reply hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala]: Another idea: what if we make metrics actor maintain his own state var.
We can create a single thread executor inside the metrics actor and submit the `Runnable` like this to that executor:
```
val futureRes =  countRootWorkflowSummaryEntriesByArchiveStatusAndOlderThanTimestamp()
futureRes onComplete {
  case Success =>
    sendGauge()
    self ! FinishedMetricsStuffAndCanAcceptNewRequests
    ...
}
```
If actor's state is `busy` then it will dismiss new requests from deletion actor.
After receiving `FinishedMetricsStuffAndCanAcceptNewRequests` message the metrics actor will update it's state to `free` and will accept new requests from deletion actor again.

@cjllanwarne, @mcovarr how does that sound?

[the remark this answers is not in the document]
Comment (mcovarr) [reply hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala]: Wouldn't that break the state testing below?

I guess I don't understand why the standard FSM pattern wouldn't work well here.

[the remark this answers is not in the document]
Comment (grsterin) [reply hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/DeleteMetadataActor.scala]: FSM would work. I just forgot that `countRootWorkflowSummaryEntriesByArchiveStatusAndOlderThanTimestamp()` is already asynchronous and returns `Future`.

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActor.scala [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import cats.data.NonEmptyList
+import cromwell.core.instrumentation.InstrumentationPrefixes
+import cromwell.services.MetadataServicesStore
+import cromwell.services.instrumentation.CromwellInstrumentation
+import cromwell.services.metadata.MetadataArchiveStatus
+import cromwell.services.metadata.MetadataArchiveStatus.Archived
+import cromwell.services.metadata.hybridcarbonite.NumOfWorkflowsToDeleteMetadataMetricActor._
+import cromwell.services.metadata.impl.{MetadataDatabaseAccess, MetadataServiceActor}
+
+class NumOfWorkflowsToDeleteMetadataMetricActor(override val serviceRegistryActor: ActorRef)
```
Comment (cjllanwarne) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActor.scala]: Since it's already a long name, might as well make it `NumberOfWorkflows[...]`?

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActor.scala [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        MetadataArchiveStatus.toDatabaseValue(Archived),
+        currentTimestampMinusDelay
+      )
+        .map(intVal => FinishedNumOfWorkflowsToDeleteMetadataMetricValueCalculation(intVal.toLong))
+        .pipeTo(self)
+      goto(MetricCalculationInProgress)
+  }
+
+  when(MetricCalculationInProgress) {
+    case Event(CalculateNumOfWorkflowsToDeleteMetadataMetricValue, _) =>
```
Comment (cjllanwarne) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActor.scala]: You also need to handle `NumOfWorkflowsToDeleteMetadataMetricValue` in this state. 

Since it's kind of an edge case that we're still calculating a query when we get an exact number from the delete actor, I wouldn't try to be too clever about handling it (eg by cancelling our own query or changing states before our query responds). I'd probably send the new metric to the instrumentation actor but otherwise stay in the same state and wait for our existing query to give us its answer.

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActor.scala [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    MetadataServiceActor.MetadataInstrumentationPrefix :+ "delete" :+ "numOfWorkflowsToDeleteMetadata"
+
+  // states
+  sealed trait MetadataDeletionMetricsActorState
+  case object WaitingForMetricCalculationRequestOrMetricValue extends MetadataDeletionMetricsActorState
+  case object MetricCalculationInProgress extends MetadataDeletionMetricsActorState
+
+  // messages
+  case class CalculateNumOfWorkflowsToDeleteMetadataMetricValue(currentTimestampMinusDelay: OffsetDateTime)
+  case class NumOfWorkflowsToDeleteMetadataMetricValue(value: Long)
```
Comment (cjllanwarne) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActor.scala]: Also suggest making the `Number[...]`

File: hybridCarboniteMetadataService/src/test/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActorSpec.scala [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    val dbFailurePromise = Promise[Int]()
+    val metricActor = TestFSMRef(new NumOfWorkflowsToDeleteMetadataMetricActor(TestProbe().ref) {
+      override def countRootWorkflowSummaryEntriesByArchiveStatusAndOlderThanTimestamp(archiveStatus: Option[String], thresholdTimestamp: OffsetDateTime)(implicit ec: ExecutionContext): Future[Int] =
+        dbFailurePromise.future
+    })
+    metricActor ! CalculateNumOfWorkflowsToDeleteMetadataMetricValue(OffsetDateTime.now())
+    eventually {
+      metricActor.stateName shouldBe MetricCalculationInProgress
+    }
+    dbFailurePromise.failure(new RuntimeException("TEST: cannot query data from DB"))
```
Comment (cjllanwarne) [inline hybridCarboniteMetadataService/src/test/scala/cromwell/services/metadata/hybridcarbonite/NumOfWorkflowsToDeleteMetadataMetricActorSpec.scala]: Nice! I like this test a lot! 💯

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumberOfWorkflowsToDeleteMetadataMetricActor.scala [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  when(MetricCalculationInProgress) {
+    case Event(CalculateNumberOfWorkflowsToDeleteMetadataMetricValue, _) =>
+      // metric actor is already busy calculating metric value, so we dismiss this request
+      stay()
+    case Event(NumberOfWorkflowsToDeleteMetadataMetricValue(value), _) =>
+      // pass through the pre-calculated metric value and stay, continuing to wait for DB query to finish
+      sendGauge(workflowsToDeleteMetadataMetricPath, value, InstrumentationPrefixes.ServicesPrefix)
+      stay()
```
Comment (mcovarr) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumberOfWorkflowsToDeleteMetadataMetricActor.scala]: TOL Is this message expected in this state? If this really should be the same code in both states could it be DRYed out to `whenUnhandled` (or is the comment here important enough to keep the duplication)?

Comment (cjllanwarne) [reply hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumberOfWorkflowsToDeleteMetadataMetricActor.scala]: Probably not "expected" but certainly possible. We should do something sensible if and when it happens.

Comment (grsterin) [reply hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/NumberOfWorkflowsToDeleteMetadataMetricActor.scala]: DRYed
